File: Actividad 4/Actividad 4.1/4_4.cpp
#include <iostream>
#include <vector>
using namespace std;


class Tienda{
    public:
        string nombre;
        string direccion;
        string tipo;
        vector<string> productos_disponibles;
        vector<float> precios;
        vector<string> compras;
        Tienda(string nombre, string direccion, string tipo){
            this->nombre = nombre;
            this->direccion = direccion;
            this->tipo = tipo;
        }

        void generar_lista_productos(){
            string producto;
            float precio;
            cout<<"Ingrese el nombre del producto: ";
            cin>>producto;
            cout<<"Ingrese el precio del producto: ";
            cin>>precio;
            productos_disponibles.push_back(producto);
            precios.push_back(precio);
        }
        void ver_lista_productos_disponibles(){
            for(int i = 0; i < productos_disponibles.size(); i++){
                cout<<productos_disponibles[i]<<" - "<<precios[i]<<endl;
            }
        }

        void ver_lista_de_compras(){
            for(int i = 0; i < compras.size(); i++){
                cout<<compras[i]<<endl;
            }
        }

};

class Cliente {
    public:
        string nombres;
        string apellidos;
        string direccion;
        float cuenta;
        vector<string> lista_compras;
        Cliente(string nombres, string apellidos, string direccion){
            this->nombres = nombres;
            this->apellidos = apellidos;
            this->direccion = direccion;

        }
    
        void agregar_productos_lista(){
            string producto;
            cout<<"¿Que producto desea agregar? " ;
            cin>>producto;
            lista_compras.push_back(producto);
        }

        void pagar_cuenta(Tienda tienda){
            float total = 0;
            float pago = 0;
            for(int i = 0; i < lista_compras.size(); i++){
                for(int j = 0; j < tienda.productos_disponibles.size(); j++){
                    if(lista_compras[i] == tienda.productos_disponibles[j]){
                        total += tienda.precios[j];
                    }
                }
            }
            cout << "El precio total es: " << total << endl;
            cout << "¿Con cuanto pagará su cuenta? ";
            cin >> pago;
            if(pago >= total){
                cout << "Su vuelto es: " << pago - total << endl;
                tienda.compras.insert(tienda.compras.end(), lista_compras.begin(), lista_compras.end());
            } else {
                cout << "No tiene suficiente dinero" << endl;
            }
            
        }
};

int menu(){
    int opcion;
    cout<<"Menu"<<endl;
    cout<<"1. Agregar productos a la tienda"<<endl;
    cout<<"2. Ver lista de productos disponibles"<<endl;
    cout<<"3. Agregar productos a la lista de compras"<<endl;
    cout<<"4. Ver lista de compras"<<endl;
    cout<<"5. Calcular total y pagar"<<endl;
    cout<<"6. Salir"<<endl;
    cout<<"Elija una opción: ";
    cin>>opcion;
    return opcion;
}

int main(){
    vector<Tienda> tiendas;
    vector<Cliente> clientes;
    string nombre, direccion, tipo, nombres, apellidos, direccion_cliente;
    int opcion;
    while(true){
        cout<<"Nombre de la tienda: ";
        cin>>nombre;
        cout<<"Direccion de la tienda: ";
        cin>>direccion;
        cout<<"Tipo de tienda: ";
        cin>>tipo;
        Tienda tienda(nombre, direccion, tipo);
        tiendas.push_back(tienda);
        cout << endl;
        cout<<"Nombres del cliente: ";
        cin>>nombres;
        cout<<"Apellidos del cliente: ";
        cin>>apellidos;
        cout<<"Direccion del cliente: ";
        cin>>direccion_cliente;
        Cliente cliente(nombres, apellidos, direccion_cliente);
        clientes.push_back(cliente);
        cout << endl;
        while(true){
            cout<<"Ingrese el nombre de la tienda: ";
            cin>>nombre;
            for(int i = 0; i < tiendas.size(); i++){
                if(tiendas[i].nombre == nombre){
                    cout<<"Ingrese el nombre del cliente: ";
                    cin>>nombres;
                    cout<< endl;
                    for(int j = 0; j < clientes.size(); j++){
                        if(clientes[j].nombres == nombres){
                            while(true){
                                opcion = menu();
                                if(opcion == 1){
                                    tiendas[i].generar_lista_productos();
                                } else if(opcion == 2){
                                    tiendas[i].ver_lista_productos_disponibles();
                                } else if(opcion == 3){
                                    clientes[j].agregar_productos_lista();
                                } else if(opcion == 4){
                                    tiendas[i].ver_lista_de_compras();
                                } else if(opcion == 5){
                                    clientes[j].pagar_cuenta(tiendas[i]);
                                } else if(opcion == 6){
                                    return 0;
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: Actividad 4/Actividad 4.1/4_4.java
import java.util.ArrayList;
import java.util.Scanner;

class Tienda {
    String nombre;
    String direccion;
    String tipo;
    ArrayList<String> productos_disponibles = new ArrayList<>();
    ArrayList<Float> precios = new ArrayList<>();
    ArrayList<String> compras = new ArrayList<>();

    public Tienda(String nombre, String direccion, String tipo) {
        this.nombre = nombre;
        this.direccion = direccion;
        this.tipo = tipo;
    }

    public void generarListaProductos() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Ingrese el nombre del producto: ");
        String producto = scanner.next();
        System.out.print("Ingrese el precio del producto: ");
        float precio = scanner.nextFloat();
        productos_disponibles.add(producto);
        precios.add(precio);
    }

    public void verListaProductosDisponibles() {
        for (int i = 0; i < productos_disponibles.size(); i++) {
            System.out.println(productos_disponibles.get(i) + " - " + precios.get(i));
        }
    }

    public void verListaDeCompras() {
        for (String compra : compras) {
            System.out.println(compra);
        }
    }
}

class Cliente {
    String nombres;
    String apellidos;
    String direccion;
    ArrayList<String> lista_compras = new ArrayList<>();

    public Cliente(String nombres, String apellidos, String direccion) {
        this.nombres = nombres;
        this.apellidos = apellidos;
        this.direccion = direccion;
    }

    public void agregarProductosLista() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("¿Qué producto desea agregar? ");
        String producto = scanner.next();
        lista_compras.add(producto);
    }

    public void pagarCuenta(Tienda tienda) {
        float total = 0;
        float pago = 0;
        for (String compra : lista_compras) {
            for (int i = 0; i < tienda.productos_disponibles.size(); i++) {
                if (compra.equals(tienda.productos_disponibles.get(i))) {
                    total += tienda.precios.get(i);
                }
            }
        }
        Scanner scanner = new Scanner(System.in);
        System.out.println("El precio total es: " + total);
        System.out.print("¿Con cuanto pagará su cuenta? ");
        pago = scanner.nextFloat();
        if (pago >= total) {
            System.out.println("Su vuelto es: " + (pago - total));
            tienda.compras.addAll(lista_compras);
        } else {
            System.out.println("No tiene suficiente dinero");
        }
    }
}

public class Main {
    static Scanner scanner = new Scanner(System.in);
    static ArrayList<Tienda> tiendas = new ArrayList<>();
    static ArrayList<Cliente> clientes = new ArrayList<>();

    public static void main(String[] args) {
        while (true) {
            String nombre = ingresarDatos("Nombre de la tienda: ");
            String direccion = ingresarDatos("Direccion de la tienda: ");
            String tipo = ingresarDatos("Tipo de tienda: ");
            Tienda tienda = new Tienda(nombre, direccion, tipo);
            tiendas.add(tienda);
            System.out.println();
            nombre = ingresarDatos("Nombres del cliente: ");
            String apellidos = ingresarDatos("Apellidos del cliente: ");
            String direccionCliente = ingresarDatos("Direccion del cliente: ");
            Cliente cliente = new Cliente(nombre, apellidos, direccionCliente);
            clientes.add(cliente);
            System.out.println();
            while (true) {
                nombre = ingresarDatos("Ingrese el nombre de la tienda: ");
                for (Tienda tienda1 : tiendas) {
                    if (tienda1.nombre.equals(nombre)) {
                        nombre = ingresarDatos("Ingrese el nombre del cliente: ");
                        System.out.println();
                        for (Cliente cliente1 : clientes) {
                            if (cliente1.nombres.equals(nombre)) {
                                while (true) {
                                    int opcion = menu();
                                    if (opcion == 1) {
                                        tienda1.generarListaProductos();
                                    } else if (opcion == 2) {
                                        tienda1.verListaProductosDisponibles();
                                    } else if (opcion == 3) {
                                        cliente1.agregarProductosLista();
                                    } else if (opcion == 4) {
                                        tienda1.verListaDeCompras();
                                    } else if (opcion == 5) {
                                        cliente1.pagarCuenta(tienda1);
                                    } else if (opcion == 6) {
                                        return;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    static String ingresarDatos(String mensaje) {
        System.out.print(mensaje);
        return scanner.next();
    }

    static int menu() {
        System.out.println("Menu");
        System.out.println("1. Agregar productos a la tienda");
        System.out.println("2. Ver lista de productos disponibles");
        System.out.println("3. Agregar productos a la lista de compras");
        System.out.println("4. Ver lista de compras");
        System.out.println("5. Calcular total y pagar");
        System.out.println("6. Salir");
        System.out.print("Elija una opción: ");
        return scanner.nextInt();
    }
}
